File: frontend/types/websocket.ts
import type { StructuredError } from "@/types/error";
import type { CurrentMarketOrderbook, Market, OrderbookSnapshot } from "@/types/market";
import type { PnlSummary } from "@/types/pnl";

export type MarketTick = {
  token_id: string;
  event_type: string | null;
  best_bid: string | null;
  best_ask: string | null;
  midpoint: string | null;
  spread: string | null;
  last_trade_price: string | null;
  data_source: string;
  raw_payload: Record<string, unknown>;
  received_at: string;
};

export type BtcPriceTick = {
  symbol: string;
  value: string;
  source: string;
  source_timestamp: string | null;
  received_at: string;
  raw_payload: Record<string, unknown>;
};

export type StrategyDecision = Record<string, unknown>;
export type OrderUpdate = Record<string, unknown>;

export type BotStatus = {
  running: boolean;
  market_ws_fresh?: boolean;
  rtds_fresh?: boolean;
  background_workers_required?: boolean;
  message?: string | null;
};

export type RiskStatus = {
  trading_enabled: boolean;
  kill_switch_active: boolean;
  geoblock_blocked?: boolean | null;
};

export type RuntimeStatus = {
  status: string;
  timestamp?: string;
  reconnect_attempts?: number;
  last_tick_received_at?: string | null;
  subscribed_symbol?: string;
  asset_ids?: string[];
  message?: string;
};

export type DashboardWsEvent =
  | { type: "current_market"; data: Market }
  | { type: "orderbook_update"; data: CurrentMarketOrderbook }
  | { type: "orderbook_snapshot"; data: OrderbookSnapshot & { data_source?: string } }
  | { type: "market_tick"; data: MarketTick }
  | { type: "btc_price_tick"; data: BtcPriceTick }
  | { type: "rtds_status"; data: RuntimeStatus }
  | { type: "market_ws_status"; data: RuntimeStatus }
  | { type: "trade_tick"; data: MarketTick }
  | { type: "strategy_decision"; data: StrategyDecision }
  | { type: "order_update"; data: OrderUpdate }
  | { type: "bot_status"; data: BotStatus }
  | { type: "risk_status"; data: RiskStatus }
  | { type: "pnl_summary"; data: PnlSummary }
  | { type: "error"; data: StructuredError };

File: backend/app/schemas/websocket.py
from datetime import datetime
from decimal import Decimal
from typing import Any, Literal

from pydantic import BaseModel, Field


class MarketTick(BaseModel):
    token_id: str
    event_type: str | None = None
    best_bid: Decimal | None = None
    best_ask: Decimal | None = None
    midpoint: Decimal | None = None
    spread: Decimal | None = None
    last_trade_price: Decimal | None = None
    data_source: str = "market_ws"
    raw_payload: dict[str, Any]
    received_at: datetime


class BtcPriceTick(BaseModel):
    symbol: str = "btc/usd"
    value: Decimal
    source: str = "polymarket_rtds_chainlink"
    source_timestamp: datetime | None = None
    received_at: datetime
    raw_payload: dict[str, Any]


class BotStatus(BaseModel):
    running: bool = False
    market_ws_fresh: bool = False
    rtds_fresh: bool = False
    background_workers_required: bool = True
    message: str | None = None


class RiskStatus(BaseModel):
    trading_enabled: bool = False
    kill_switch_active: bool = False
    geoblock_blocked: bool | None = None


class DashboardError(BaseModel):
    code: str
    title: str | None = None
    message: str


class DashboardWsEvent(BaseModel):
    type: Literal[
        "market_tick",
        "btc_price_tick",
        "strategy_decision",
        "order_update",
        "bot_status",
        "risk_status",
        "pnl_summary",
        "current_market",
        "orderbook_update",
        "orderbook_snapshot",
        "rtds_status",
        "market_ws_status",
        "trade_tick",
        "error",
    ]
    data: dict[str, Any] = Field(default_factory=dict)
